kafkabp/consumer.go

```diff
@@ -0,0 +1,229 @@
+		if err != nil {
+			return err
+		}
+
+		kc.consumer.Store(c)
+		kc.partitions.Store(partitions)
+		return nil
+	}
+
+	err := rebalance()
+	if err != nil {
+		metricsbp.M.Counter("kafka.consumer.rebalance.failure").Add(1)
+		return err
+	}
+
+	metricsbp.M.Counter("kafka.consumer.rebalance.success").Add(1)
+	return nil
+}
+
+// Close closes all partition consumers first, then the parent consumer.
+func (kc *consumer) Close() error {
+	// Return early if closing is already in progress
+	if !atomic.CompareAndSwapInt64(&kc.closed, 0, 1) {
+		return nil
+	}
+
+	partitionConsumers := kc.getPartitionConsumers()
+	for _, pc := range partitionConsumers {
+		// leaves room to drain pc's message and error channels
+		pc.AsyncClose()
+	}
+	// wait for the Consume function to return
+	kc.wg.Wait()
+	return kc.getConsumer().Close()
+}
+
+// Consume consumes Kafka messages and errors from each partition's consumer.
+// It is necessary to call Close() on the KafkaConsumer instance once all
+// operations are done with the consumer instance.
+func (kc *consumer) Consume(
+	messagesFunc ConsumeMessageFunc,
+	errorsFunc ConsumeErrorFunc,
+) error {
+	defer atomic.StoreInt64(&kc.consumeReturned, 1)
+	kc.wg.Add(1)
+	defer kc.wg.Done()
+
+	// sarama could close the channels (and cause the goroutines to finish) in
+	// two cases, where we want different behavior:
+	//   - in case of partition rebalance: restart goroutines
+	//   - in case of call to Close/AsyncClose: exit
+	var wg sync.WaitGroup
+	for {
+		// create a partition consumer for each partition
+		consumer := kc.getConsumer()
+		partitions := kc.getPartitions()
+		partitionConsumers := make([]sarama.PartitionConsumer, 0, len(partitions))
+
+		for _, p := range partitions {
+			partitionConsumer, err := consumer.ConsumePartition(kc.cfg.Topic, p, kc.cfg.Offset)
+			if err != nil {
+				return err
+			}
+			partitionConsumers = append(partitionConsumers, partitionConsumer) // for closing individual partitions when Close() is called
+
+			// consume partition consumer messages
+			wg.Add(1)
+			go func(pc sarama.PartitionConsumer) {
+				defer wg.Done()
+				for m := range pc.Messages() {
+					// Wrap in anonymous function for easier defer.
+					func() {
+						ctx := context.Background()
+						var err error
+						var span *tracing.Span
+						spanName := "consumer." + kc.cfg.Topic
+						ctx, span = tracing.StartTopLevelServerSpan(ctx, spanName)
+						defer func() {
+							span.FinishWithOptions(tracing.FinishOptions{
+								Ctx: ctx,
+								Err: err,
+							}.Convert())
+						}()
+
+						err = messagesFunc(ctx, m)
+					}()
+				}
+			}(partitionConsumer)
+
+			wg.Add(1)
+			// consume partition consumer errors
+			go func(pc sarama.PartitionConsumer) {
+				defer wg.Done()
+				for err := range pc.Errors() {
+					errorsFunc(err)
+				}
+			}(partitionConsumer)
+		}
+		kc.partitionConsumers.Store(partitionConsumers)
+
+		wg.Wait()
+
+		if atomic.LoadInt64(&kc.closed) != 0 {
```

Comment:
maybe add a comment to the effect of "close was called, shutting down the consumer"